src/notation/view/notationnavigator.cpp

```diff
@@ -0,0 +1,314 @@
+{
+    if (!rect.isValid()) {
+        return;
+    }
+
+    moveCanvasToRect(rect);
+
+    m_viewRect = rect;
+    rescale();
+    update();
+}
+
+int NotationNavigator::orientation() const
+{
+    return static_cast<int>(configuration()->navigatorOrientation());
+}
+
+void NotationNavigator::onCurrentNotationChanged()
+{
+    auto notation = currentNotation();
+    setVisible(notation != nullptr);
+
+    if (!notation) {
+        return;
+    }
+
+    notation->notationChanged().onNotify(this, [this]() {
+        rescale();
+        update();
+    });
+}
+
+INotationPtr NotationNavigator::currentNotation() const
+{
+    return globalContext()->currentNotation();
+}
+
+ViewMode NotationNavigator::notationViewMode() const
+{
+    auto notation = currentNotation();
+    if (!notation) {
+        return ViewMode::PAGE;
+    }
+
+    return notation->viewMode();
+}
+
+void NotationNavigator::paint(QPainter* painter)
+{
+    painter->fillRect(viewport(), configuration()->backgroundColor());
+
+    paintPages(painter);
+    paintViewRect(painter);
+}
+
+void NotationNavigator::paintPages(QPainter* painter)
+{
+    PageList pages = this->pages();
+    if (pages.empty()) {
+        return;
+    }
+
+    ViewMode viewMode = notationViewMode();
+    QFont font(QString::fromStdString(configuration()->fontFamily()), 2000);
+
+    painter->setTransform(m_matrix);
+    QRectF viewportRect = m_matrix.inverted().mapRect(QRectF(viewport()));
+
+    for (const Page* page : pages) {
+        QPointF pos(page->pos());
+        QRectF pageRect(page->abbox().translated(pos));
+        if (pageRect.right() < viewportRect.left()) {
+            continue;
+        }
+        if (pageRect.left() > viewportRect.right()) {
+            break;
+        }
+
+        painter->translate(pos);
+
+        painter->fillRect(page->bbox(), configuration()->pageColor());
+
+        //! todo
+        Page* _page = const_cast<Page*>(page);
+        paintElements(painter, _page->elements());
+
+        if (viewMode == ViewMode::PAGE) {
+            painter->setFont(font);
+            painter->setPen(configuration()->layoutBreakColor());
+            painter->drawText(page->bbox(), Qt::AlignCenter, QString("%1").arg(page->no() + 1));
+        }
+        painter->translate(-pos);
+    }
+}
+
+void NotationNavigator::paintElements(QPainter* painter, const QList<Element*>& elements)
+{
+    for (const Ms::Element* element : elements) {
+        if (!element->visible()) {
+            continue;
+        }
+
+        element->itemDiscovered = false;
+        QPointF elementPosition(element->pagePos());
+
+        painter->translate(elementPosition);
+        element->draw(painter);
+        painter->translate(-elementPosition);
+    }
+}
+
+void NotationNavigator::paintViewRect(QPainter* painter)
+{
+    QColor color(configuration()->selectionColor());
+    QPen pen(color, 2.0);
+    painter->setPen(pen);
+    painter->setBrush(QColor(color.red(), color.green(), color.blue(), 40));
```

Comment:
Please, add named constants (or methods in configuration) to all magic numbers in this method